File: app/src/main/java/br/com/pedrosilva/tecnonutri/presentation/presenters/impl/FeedItemPresenterImpl.java
package br.com.pedrosilva.tecnonutri.presentation.presenters.impl;

import br.com.pedrosilva.tecnonutri.domain.entities.FeedItem;
import br.com.pedrosilva.tecnonutri.domain.executor.Executor;
import br.com.pedrosilva.tecnonutri.domain.executor.MainThread;
import br.com.pedrosilva.tecnonutri.domain.interactors.ChangeLikeInteractor;
import br.com.pedrosilva.tecnonutri.domain.interactors.GetFeedItemInteractor;
import br.com.pedrosilva.tecnonutri.domain.interactors.base.Interactor;
import br.com.pedrosilva.tecnonutri.domain.interactors.impl.ChangeLikeInteractorImpl;
import br.com.pedrosilva.tecnonutri.domain.interactors.impl.GetFeedItemInteractorImpl;
import br.com.pedrosilva.tecnonutri.domain.repositories.FeedRepository;
import br.com.pedrosilva.tecnonutri.presentation.presenters.FeedItemPresenter;
import br.com.pedrosilva.tecnonutri.presentation.presenters.base.AbstractPresenter;

/**
 * Created by psilva on 3/20/17.
 */

public class FeedItemPresenterImpl extends AbstractPresenter implements FeedItemPresenter, GetFeedItemInteractor.Callback, ChangeLikeInteractor.Callback {
    private final View view;
    private final FeedRepository feedRepository;
    private final String feedHash;

    public FeedItemPresenterImpl(Executor executor,
                                 MainThread mainThread,
                                 View view, FeedRepository feedRepository, String feedHash) {
        super(executor, mainThread);

        this.view = view;
        this.feedRepository = feedRepository;
        this.feedHash = feedHash;

        load();
    }

    public void load() {
        //view.showProgress();
        Interactor interactor = new GetFeedItemInteractorImpl(
                executor,
                mainThread,
                feedRepository,
                feedHash,
                this
        );
        interactor.execute();
    }

    @Override
    public void resume() {

    }

    @Override
    public void pause() {

    }

    @Override
    public void stop() {

    }

    @Override
    public void destroy() {

    }

    @Override
    public void onError(String message) {

    }

    @Override
    public void onFeedItemRetrieved(FeedItem feedItem) {
        view.hideProgress();
        view.onLoadFeedItem(feedItem);
    }

    @Override
    public void onFeedItemRetrieveFailed(Throwable t) {
        view.onLoadFail(t);
    }

    @Override
    public void reload() {
        load();
    }

    @Override
    public void changeLike(String feedHash, boolean liked) {
        Interactor interactor = new ChangeLikeInteractorImpl(
                executor,
                mainThread,
                feedRepository,
                feedHash,
                liked,
                this
        );
        interactor.execute();
    }

    @Override
    public boolean isLiked(String feedHash) {
        return feedRepository.isLiked(feedHash);
    }

    @Override
    public void onChange(String feedHash, boolean liked) {
        view.onChangeLike(feedHash, liked);
    }
}

File: app/src/main/kotlin/br/com/pedrosilva/tecnonutri/presentation/presenters/impl/FeedItemPresenterImpl.kt
package br.com.pedrosilva.tecnonutri.presentation.presenters.impl

import com.pedrenrique.tecnonutri.domain.FeedItem
import com.pedrenrique.tecnonutri.domain.executor.Executor
import com.pedrenrique.tecnonutri.domain.executor.MainThread
import com.pedrenrique.tecnonutri.domain.interactors.ChangeLikeInteractor
import com.pedrenrique.tecnonutri.domain.interactors.GetFeedItemInteractor
import com.pedrenrique.tecnonutri.domain.interactors.base.Interactor
import com.pedrenrique.tecnonutri.domain.interactors.impl.ChangeLikeInteractorImpl
import com.pedrenrique.tecnonutri.domain.interactors.impl.GetFeedItemInteractorImpl
import com.pedrenrique.tecnonutri.domain.repositories.FeedRepository
import br.com.pedrosilva.tecnonutri.presentation.presenters.FeedItemPresenter
import br.com.pedrosilva.tecnonutri.presentation.presenters.base.AbstractPresenter

class FeedItemPresenterImpl(
    executor: Executor,
    mainThread: MainThread,
    private val view: FeedItemPresenter.View,
    private val feedRepository: FeedRepository,
    private val feedHash: String
) : AbstractPresenter(executor, mainThread),
    FeedItemPresenter,
    GetFeedItemInteractor.Callback,
    ChangeLikeInteractor.Callback {

    init {
        load()
    }

    fun load() {
        //view.showProgress();
        val interactor: Interactor = GetFeedItemInteractorImpl(
            executor,
            mainThread,
            feedRepository,
            feedHash,
            this
        )
        interactor.execute()
    }

    override fun resume() {}

    override fun pause() {}

    override fun stop() {}

    override fun destroy() {}

    override fun onError(message: String) {}

    override fun onFeedItemRetrieved(feedItem: FeedItem) {
        view.hideProgress()
        view.onLoadFeedItem(feedItem)
    }

    override fun onFeedItemRetrieveFailed(error: Throwable) {
        view.onLoadFail(error)
    }

    override fun reload() {
        load()
    }

    override fun changeLike(feedHash: String, liked: Boolean) {
        val interactor: Interactor = ChangeLikeInteractorImpl(
            executor,
            mainThread,
            feedRepository,
            feedHash,
            liked,
            this
        )
        interactor.execute()
    }

    override fun isLiked(feedHash: String): Boolean {
        return feedRepository.isLiked(feedHash)
    }

    override fun onChange(feedHash: String, liked: Boolean) {
        view.onChangeLike(feedHash, liked)
    }
}